core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,399 @@
+    this.verification = verification;
+    this.peerVerifier = peerVerifier;
+    this.sslSocketPeerVerifier = socketPeerVerifier;
+    this.sslEnginePeerVerifier = enginePeerVerifier;
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, null, false);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, socket, false);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+      throws CertificateException {
+    checkTrusted(chain, authType, engine, null, false);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain,  String authType, SSLEngine engine)
+      throws CertificateException {
+    checkTrusted(chain, authType, engine, null, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, null, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, socket, true);
+  }
+
+  @Override
+  public X509Certificate[] getAcceptedIssuers() {
+    return trustCerts;
+  }
+
+  // If this is set, we will use the default trust certificates stored on user's local system.
+  // After this is used, functions that will update this.trustCerts(e.g. updateTrustCredentials(),
+  // updateTrustCredentialsFromFile()) should not be called.
+  public void useSystemDefaultTrustCerts() {
+    this.trustCerts = new X509Certificate[0];
+  }
+
+  /**
+   * Updates the current cached trust certificates.
+   *
+   * @param trustCerts  the trust certificates that are going to be used
+   */
+  public void updateTrustCredentials(X509Certificate[] trustCerts) {
+    // Question: in design, we will do an copy of trustCerts, but that seems not easy, since
+    // X509Certificate is an abstract class without copy constructor.
+    // Can we document in the API that trustCerts shouldn't be modified once set?
+    this.trustCerts = trustCerts;
+  }
+
+  private void checkTrusted(X509Certificate[] chain, String authType, SSLEngine sslEngine,
+      Socket socket, boolean checkingServer) throws CertificateException {
+    if (this.verification == Verification.CertificateAndHostNameVerification
+        || this.verification == Verification.CertificateOnlyVerification) {
+      if (chain == null || chain.length == 0) {
+        throw new CertificateException(
+            "Want certificate verification but got null or empty certificates");
+      }
+      // Set the cached trust certificates into a key store.
+      KeyStore ks;
```

Comment:
We should not do all of this creation per-check. We should do it up-front and reuse it.

Reply:
Done. Move the creation of the `KeyStore` to the constructor, and added one `ks` member to the class.

Reply:
Do you mean the `KeyManager`? I added a `KeyStore` in the `KeyManager` to keep consistent with `TrustManager`